course7/course_en.md

```diff
@@ -0,0 +1,108 @@
+# Modern Programming Thoughts: Imperative Programming
+
+## Revisiting Functional Programming
+
+The core idea of functional programming is that each input corresponds to a fixed output, and identifiers can be directly substituted with their corresponding values, a property known as referential transparency. However, developing practical programs often requires "side effects" beyond calculations, such as input/output operations and modifying data in memory. These side effects may lead to inconsistent results upon multiple executions, breaking referential transparency.
+
+```moonbit
+let x: Int = 1 + 1 // x can be directly replaced with 2
+fn square(x: Int) -> Int { x * x }
+let z: Int = square(x) // can be replaced with { 2 * 2 }, still resulting in 4
+```
+
+## Commands and Side Effects
+
+The `print` command in MoonBit enables us to output strings, such as `print("hello moonbit")`. We can define initialization instructions in the `init` code block, serving as the program entry point. However, output commands may break referential transparency.
+
+```moonbit
+fn init {
+  let x: Int = {
+    println("hello moonbit") // First, execute the command and perform output
+    1 + 1 // Then, take the last value of the expression block as the block's value
+  }
+  let z: Int = square(x) // Output once
+}
+```
+
+We typically use the unit type `Unit` to represent the return value of functions or commands with side effects. It has only one value: `()`.
+
+## Variables and Aliases
+
+In MoonBit, we can define temporary variables using `let mut`. The assignment operation itself is a command. Struct fields are immutable by default, but we can use `mut` to identify mutable fields, which are treated as references.
+
+```moonbit
+struct Ref[T] { mut val : T }
+
+fn init {
+  let ref: Ref[Int] = { val: 1 } // ref itself is just a data binding
+  ref.val = 10 // We can modify the fields of the struct
+  println(ref.val.to_string()) // Output 10
+}
+```
+
+Multiple identifiers pointing to the same mutable data structure can be considered aliases, which need to be handled carefully.
+
+## Loops
+
+We can define loops using variables in MoonBit. A loop includes defining the loop variable and its initial value, checking whether to continue the loop, and iterating the variable.
+
+```moonbit
+let mut i = 0
+while i < 2, i = i + 1 {
+  println("Output")
+} // Repeat output 2 times
+```
+
+The loop execution flow is: check the condition -> execute commands -> iterate the variable -> repeat the above process. MoonBit provides a debugger that enables observing the changes in runtime data during the loop process.
+
+## Loops and Recursion
+
+Loops and recursion are equivalent. A loop can be written in a recursive form. Taking the calculation of the Fibonacci sequence as an example:
+
+```moonbit
+// Loop form
+fn fib_mut(n: Int) -> Int {
+  let mut a = 0; let mut b = 1; let mut i = 0
```

Comment:
Since this is a document not a presentation, you may write them in multiple lines.